In [2]:
import csv
import numpy as np

In [3]:
# List Comprehension
# Get all integers out of a list of numbers, e.x. [1,2.5,100.3,3.0,-1.5,-5]
# Find all numbers from 1-1000 that have a 3 in them.
# Count the number of alphabetic characters in a string (a-z, lower case or upper case)
# [Challenge - requires nesting] - Find all numbers between 1-1000 that are not divisble by any number between 2 and 9.

# SQL/Big Query (Using the noaa historic severe storms dataset).
# Find out what state has had the most historic severe storms between 1980 and 1995
# Find the average crop damage for storms in 2011
# What was the total number of direct and indirect deaths for storms in 2000? What about the total number of direct or indirect injuries?

# Big Query / Python
# Using the same dataset as above, download a CSV to parse in python of the 1000 worst storms in 2022 by magnitude when the magnitude type is "MG" or "EG" and the magnitude was at least 80
# What percent of event types were thunderstorms vs high winds?
# What state had the least number of these types of storms?  What state had the most?
# What was the mean, median, and standard deviation of the magnitude?

## List Comprehension

In [4]:
# Get all integers out of a list of numbers, e.x. [1,2.5,100.3,3.0,-1.5,-5]
listA = [1, 2, 3, 4, 5.2, 6, 3.0, 8, 9.12, 10, 11.50, 12]

newListA = [n for n in listA if n == int(n)]

print(newListA)

[1, 2, 3, 4, 6, 3.0, 8, 10, 12]


In [5]:
# Find all numbers from 1-1000 that have a 3 in them.
listOf3 = [x for x in range(0,1001) if '3' in str(x)] #cannot use "in" operator to check integer

print(listOf3)


[3, 13, 23, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 43, 53, 63, 73, 83, 93, 103, 113, 123, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 143, 153, 163, 173, 183, 193, 203, 213, 223, 230, 231, 232, 233, 234, 235, 236, 237, 238, 239, 243, 253, 263, 273, 283, 293, 300, 301, 302, 303, 304, 305, 306, 307, 308, 309, 310, 311, 312, 313, 314, 315, 316, 317, 318, 319, 320, 321, 322, 323, 324, 325, 326, 327, 328, 329, 330, 331, 332, 333, 334, 335, 336, 337, 338, 339, 340, 341, 342, 343, 344, 345, 346, 347, 348, 349, 350, 351, 352, 353, 354, 355, 356, 357, 358, 359, 360, 361, 362, 363, 364, 365, 366, 367, 368, 369, 370, 371, 372, 373, 374, 375, 376, 377, 378, 379, 380, 381, 382, 383, 384, 385, 386, 387, 388, 389, 390, 391, 392, 393, 394, 395, 396, 397, 398, 399, 403, 413, 423, 430, 431, 432, 433, 434, 435, 436, 437, 438, 439, 443, 453, 463, 473, 483, 493, 503, 513, 523, 530, 531, 532, 533, 534, 535, 536, 537, 538, 539, 543, 553, 563, 573, 583, 593, 603, 613, 623, 630, 631, 632, 633, 634, 635,

In [6]:
# Count the number of alphabetic characters in a string (a-z, lower case or upper case)
randomString = "Please count the number of alphabetic characters in this string"

total = sum([1 for n in randomString if n.isalpha()])
print(total)



54


In [7]:
# [Challenge - requires nesting] - Find all numbers between 1-1000 that are not divisble by any number between 2 and 9.
resultList = [x for x in range(0,1001) if all([x % n != 0 for n in range(2,10)])]

print(resultList)

[1, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 121, 127, 131, 137, 139, 143, 149, 151, 157, 163, 167, 169, 173, 179, 181, 187, 191, 193, 197, 199, 209, 211, 221, 223, 227, 229, 233, 239, 241, 247, 251, 253, 257, 263, 269, 271, 277, 281, 283, 289, 293, 299, 307, 311, 313, 317, 319, 323, 331, 337, 341, 347, 349, 353, 359, 361, 367, 373, 377, 379, 383, 389, 391, 397, 401, 403, 407, 409, 419, 421, 431, 433, 437, 439, 443, 449, 451, 457, 461, 463, 467, 473, 479, 481, 487, 491, 493, 499, 503, 509, 517, 521, 523, 527, 529, 533, 541, 547, 551, 557, 559, 563, 569, 571, 577, 583, 587, 589, 593, 599, 601, 607, 611, 613, 617, 619, 629, 631, 641, 643, 647, 649, 653, 659, 661, 667, 671, 673, 677, 683, 689, 691, 697, 701, 703, 709, 713, 719, 727, 731, 733, 737, 739, 743, 751, 757, 761, 767, 769, 773, 779, 781, 787, 793, 797, 799, 803, 809, 811, 817, 821, 823, 827, 829, 839, 841, 851, 853, 857, 859, 863, 869, 871, 877, 881, 883, 887, 89

## SQL/Big Query (Using the noaa historic severe storms dataset)

In [8]:
# Find out what state has had the most historic severe storms between 1980 and 1995

# SELECT state, COUNT(*) AS state_count
# FROM `bigquery-public-data.noaa_historic_severe_storms.storms_19*` 
# WHERE _TABLE_SUFFIX BETWEEN '80' AND '95'
# GROUP BY state
# ORDER BY state_count DESC;

# OR

# SELECT state, COUNT(*) AS state_count
# FROM (
#   SELECT state FROM `bigquery-public-data.noaa_historic_severe_storms.storms_1980`
#   UNION ALL
#   SELECT state FROM `bigquery-public-data.noaa_historic_severe_storms.storms_1981`
#   UNION ALL
#   SELECT state FROM `bigquery-public-data.noaa_historic_severe_storms.storms_1982`
# ) 
# GROUP BY state
# ORDER BY state_count DESC;

In [9]:
# Find the average crop damage for storms in 2011

#### SELECT AVG(damage_crops) as avg_crop_damage FROM table_2011;
#### Answer: 40133.495685522408

In [10]:
# What was the total number of direct and indirect deaths for storms in 2000? What about the total number of direct or indirect injuries?

#### SELECT SUM(deaths_direct + deaths_indirect) as total_deaths FROM table_2000
#### Answer: 521

#### SLECT SUM(injuries_direct + injuries_indirect) as total_injuries FROM table_2000
#### Answer: 3658

## Big Query/Python

In [11]:
# Using the same dataset as above, download a CSV to parse in python of the 1000 worst storms in 2022 by magnitude when the magnitude type is "MG" or "EG" and the magnitude was at least 80
class SevereStorm:
    def __init__(self, state, type, magnitude): 
        self.state = state
        self.type = type
        self.magnitude = magnitude
    
    def __str__(self):
        return f"{self.state} {self.type} {str(self. magnitude)}"

In [20]:
# What percent of event types were thunderstorms vs high winds?
storms = []
with open('severe_storms_2022.csv', 'r') as csv_file:
    reader = csv.reader(csv_file)
    line_count = 0
    for row in reader:
        if line_count != 0:
            rowValues = SevereStorm(row[0], row[1], float(row[2]))
            storms.append(rowValues)
        line_count += 1

# for i in storms:
#     print(i)

# print(len(storms))

thunderstorm = [x.type for x in storms if x.type == "thunderstorm wind"]
highwind = [x.type for x in storms if x.type == "high wind"]
# for i in thunderstorm:
#     print(i)

totalThunder = np.round((len(thunderstorm) / len(storms)) * 100)
print(totalThunder)

totalHighwind = np.round((len(highwind) / len(storms) * 100))
print(totalHighwind)

67.0
33.0


In [13]:
# What state had the least number of these types of storms? What state had the most?

stateThunderCount = {}
stateHighwindCount = {}
for i in storms:
    state = i.state
    if i.type == "thunderstorm wind":
        stateThunderCount[state] = stateThunderCount.get(state, 0) + 1
    elif i.type == "high wind":
        stateHighwindCount[state] = stateHighwindCount.get(state, 0) + 1


print("stateThunderCount dict:", stateThunderCount)
print("stateHighwindCount dict:", stateHighwindCount)

stateThunderCount dict: {'So': 32, 'No': 9, 'Oh': 7, 'Ka': 23, 'De': 3, 'Mi': 17, 'Pe': 4, 'Ne': 14, 'Te': 9, 'We': 2, 'In': 3, 'Ok': 9, 'Wy': 3, 'Al': 5, 'Ar': 1, 'Wi': 2, 'Vi': 3, 'Il': 1, 'Lo': 2, 'Ma': 2, 'Fl': 1}
stateHighwindCount dict: {'Ha': 1, 'Ca': 32, 'Ut': 4, 'Wy': 8, 'Ne': 15, 'Co': 5, 'Wa': 2, 'Or': 4, 'Id': 1, 'Al': 2, 'Te': 1}


In [14]:
# state with highest and lowest thunderstorms:
stateThunderLowest = [key for key, value in stateThunderCount.items() if value == min(stateThunderCount.values())]
print(f"these state have the least thunderstorms: {stateThunderLowest} with a value of {min(stateThunderCount.values())}")


stateThunderHighest = [key for key, value in stateThunderCount.items() if value == max(stateThunderCount.values())]
print(f"these state have the most thunderstorms: {stateThunderHighest} with a value of {max(stateThunderCount.values())}")

these state have the least thunderstorms: ['Ar', 'Il', 'Fl'] with a value of 1
these state have the most thunderstorms: ['So'] with a value of 32


In [15]:
# state with the highest and lowest high wind:
stateHighwindLowest = [key for key, value in stateHighwindCount.items() if value == min(stateHighwindCount.values())]
print(f"these state have the least thunderstorms: {stateHighwindLowest} with a value of {min(stateHighwindCount.values())}")

stateHighwindHighest = [key for key, value in stateHighwindCount.items() if value == max(stateHighwindCount.values())]
print(f"these state have the least thunderstorms: {stateHighwindHighest} with a value of {max(stateHighwindCount.values())}")

these state have the least thunderstorms: ['Ha', 'Id', 'Te'] with a value of 1
these state have the least thunderstorms: ['Ca'] with a value of 32


In [26]:
# What was the mean, median, and standard deviation of the magnitude?
# magnitudes = np.array([x.magnitude for x in storms])
magnitudes = [x.magnitude for x in storms]

print(magnitudes)

magnitude_mean = np.mean(magnitudes)
magnitude_median = np.median(magnitudes)
magnitude_std = np.std(magnitudes)

print(f"mean of magnitudes: {magnitude_mean}")
print(f"median of magnitudes: {magnitude_median}")
print(f"standard of deviation of magnitude: {magnitude_std}")

[144.0, 139.0, 136.0, 129.0, 113.0, 113.0, 111.0, 109.0, 104.0, 104.0, 104.0, 104.0, 104.0, 104.0, 104.0, 104.0, 103.0, 101.0, 101.0, 101.0, 100.0, 98.0, 97.0, 97.0, 96.0, 96.0, 96.0, 96.0, 96.0, 96.0, 96.0, 96.0, 96.0, 96.0, 96.0, 96.0, 96.0, 96.0, 96.0, 95.0, 95.0, 95.0, 95.0, 95.0, 94.0, 94.0, 93.0, 93.0, 92.0, 91.0, 91.0, 91.0, 91.0, 91.0, 91.0, 90.0, 90.0, 90.0, 90.0, 90.0, 89.0, 89.0, 89.0, 89.0, 89.0, 88.0, 88.0, 87.0, 87.0, 87.0, 87.0, 87.0, 87.0, 87.0, 87.0, 87.0, 87.0, 87.0, 87.0, 87.0, 87.0, 87.0, 87.0, 87.0, 87.0, 87.0, 87.0, 87.0, 87.0, 87.0, 87.0, 87.0, 87.0, 87.0, 87.0, 87.0, 87.0, 87.0, 87.0, 87.0, 87.0, 87.0, 87.0, 87.0, 87.0, 87.0, 87.0, 87.0, 87.0, 87.0, 87.0, 87.0, 87.0, 87.0, 87.0, 87.0, 87.0, 86.0, 86.0, 86.0, 86.0, 86.0, 86.0, 86.0, 86.0, 85.0, 85.0, 85.0, 85.0, 84.0, 84.0, 84.0, 83.0, 83.0, 83.0, 83.0, 83.0, 83.0, 83.0, 83.0, 83.0, 83.0, 83.0, 83.0, 83.0, 83.0, 83.0, 83.0, 83.0, 83.0, 83.0, 83.0, 83.0, 83.0, 83.0, 83.0, 83.0, 83.0, 83.0, 83.0, 83.0, 83.0, 83.0, 